### PR TITLE
cylc run/restart re-factor

### LINKED ISSUE #1842: scan hangs during suite start-up (rebuilding the suitedb)
We rebuild the public suite DB at suite restart, from the private DB.  During the DB rebuild, `cylc scan` hangs until it times out.  In a suite with quite a large DB (84MB) one our servers this can take many minutes.

We can presumably avoid this problem by doing the DB rebuild before connecting the suite to the network (and before daemonization, so it is clear what's happening - see also #1721).

### PR DESCRIPTION
**Pull request by matthewrmshin**
Move most logic from `bin/cylc-run` and `bin/cylc-restart` into `cylc.scheduler`.
Move logic from `cylc.run` to `Scheduler.start`.
Move CLI option parser for `cylc run` and `cylc restart` from various locations to new module `cylc.scheduler_cli`.
On restart, vacuum primary runtime database before configuring Pyro. Fix #1842.

### CODE REVIEW

**Comment by matthewrmshin** [comment]
This change adds to the line count of `cylc.scheduler`, but reduces the overall line count, so I am not totally unhappy. I'll do a bit more tests tomorrow before calling for reviews.

File: lib/cylc/scheduler.py [hunk 1]
```diff
@@ -158,39 +169,113 @@ def __init__(self, is_restart=False):
...
-        self.parser.add_option(
-            "-m", "--mode",
-            help="Run mode: live, simulation, or dummy; default is live.",
-            metavar="STRING", action="store", default='live', dest="run_mode")
-
-        self.parser.add_option(
-            "--reference-log",
-            help="Generate a reference log for use in reference tests.",
-            action="store_true", default=False, dest="genref")
+    def start(self):
```
Comment (matthewrmshin) [inline lib/cylc/scheduler.py]: This is the old `main` function in the old `cylc.run`.

**Comment by matthewrmshin** [comment]
@hjoliver I believe this is now good for a review.
@arjclark please review 2.

**Comment by hjoliver** [comment]
@matthewrmshin - I was on leave yesterday, and ran out of time today.  I guess @arjclark can do first review of this if he has time, else I'll get on to it on Monday.

File: lib/cylc/scheduler.py [hunk 2]
```diff
@@ -110,15 +120,50 @@ class Scheduler(object):
     SUITE_EVENT_MAIL = 'suite-event-mail'
     FS_CHECK_PERIOD = 600.0  # 600 seconds
 
-    def __init__(self, is_restart=False):
-        self.suite = None
+    def __init__(self, is_restart, options, args):
```
Comment (arjclark) [inline lib/cylc/scheduler.py]: Not keeping the default value?

Comment (matthewrmshin) [inline lib/cylc/scheduler.py]: No need, as this is always called with the full set of arguments.

**Comment by hjoliver** [comment]
Review 1 (or 2?) looks good to me.  (I haven't run the test battery yet, however...)

**Comment by hjoliver** [comment]
I wonder if we could we get the scheduler to report "suite initializing" to clients, from early in start-up (pre-vacuuming)?

**Comment by matthewrmshin** [comment]
> I wonder if we could we get the scheduler to report "suite initializing" to clients, from early in start-up (pre-vacuuming)?

I suppose we can put all the initialisation logic under the run loop (or a run loop like block), and have any time consuming logic moved to separate processes/threads.

However, I think a change like this should be a separate pull request.

**Comment by hjoliver** [comment]
Fair enough :-)

**Comment by hjoliver** [comment]
Review 1 - good, all tests passed, no problems found.

**Comment by matthewrmshin** [comment]
Branch re-based.

**Comment by matthewrmshin** [comment]
Branch re-based again.

**Comment by arjclark** [comment]
Looks OK to me. No problems from the test-battery in my environment.

**Pull request merged**